### PR TITLE
GET formula from server

### LINKED ISSUE #282: Get available formulas from backend
- [ ] API to get available formulas
- [ ] In client, display available formulas from backend

### PR DESCRIPTION
**Pull request by jj-stigell**
Removes mock formulas from frontend. Formula information in frontend is now fetched from backend.

2 new API endpoints, one for getting basic information of available formulas and one for getting detailed information based on formula ID.

closes #282

### CODE REVIEW

File: server/src/types/formulas.ts [hunk 1]
```diff
@@ -24,6 +24,13 @@ export interface GradingInput {
   params: object | null;
 }
 
+export interface FormulaPreview {
+  id: Formula;
+  name: string;
+  attributes: Array<string>;
+  codeSnippet: string;
+}
+
```
Comment (no-comma) [inline server/src/types/formulas.ts:33]: This type seems appropriate for `common`.

File: server/src/formulas/codeSnippets.ts [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+      return sum / weightSum;
+    };`
+  },
+  {
+    id: Formula.Manual,
+    name: 'Manual',
+    attributes: ['maxPoints', 'minRequiredPoints'],
+    codeSnippet: 'placeholder'
+  }
+];
```
Comment (no-comma) [inline server/src/formulas/codeSnippets.ts:27]: Rather than having these all here, we could also store this preview information in `formulas/index.ts` and include per-formula information in the same `registerFormula` call in each formula definition file.

Comment (jj-stigell) [reply server/src/formulas/codeSnippets.ts]: moved the preview info of each formula to respective formula file, these are now registered to the `formulaImplementations`

File: server/src/formulas/weightedAverage.ts [hunk 3]
```diff
@@ -38,9 +39,29 @@ async function calculateWeightedAverage(
...
+  const [sum, weightSum] = weights.reduce(
+    (acc, w, i) => {
+      acc[0] = acc[0] + nums[i] * w;
+      acc[1] = acc[1] + w;
+      return acc;
+    },
+    [0, 0]
+  );
+  return sum / weightSum;
+};`;
```
Comment (no-comma) [inline server/src/formulas/weightedAverage.ts:53]: I would copy all this to the string:
```ts
interface WeightedAverageParams {
  min: number;
  max: number;
  weight: number;
}

async function calculateWeightedAverage(
  inputs: Array<CalculationInput>
): Promise<CalculationResult> {
  let grade: number = 0;
  let status: Status = Status.Pass;

  for (const input of inputs) {
    const subResult: CalculationResult = input.subResult;
    const params: WeightedAverageParams = input.params as WeightedAverageParams;

    if (subResult.status !== Status.Pass)
      status = Status.Fail;

    grade += subResult.grade * params.weight;
  }

  return {
    grade: grade,
    status: status,
  };
}
```

File: server/src/formulas/weightedAverage.ts [hunk 4]
```diff
@@ -38,9 +39,29 @@ async function calculateWeightedAverage(
...
+
+const name: string = 'Weighted average';
+
+const attributes: Array<string> = ['maxPoints', 'minRequiredPoints', 'weight'];
```
Comment (no-comma) [inline server/src/formulas/weightedAverage.ts:57]: `maxPoints` and `minRequiredPoints` should be removed, again related to #303. But this can also be done in another PR.

File: client/src/utils/util.ts
Comment (no-comma) [inline client/src/utils/util.ts:1]: Name this `index.ts` instead?

File: server/src/routes/formula.ts [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+import { controllerDispatcher } from '../middleware/errorHandler';
+
+export const router: Router = Router();
+
+/**
+ * @swagger
+ * definitions:
+ *   FormulaData:
+ *     type: object
+ *     description: List of available formulas.
```
Comment (no-comma) [inline server/src/routes/formula.ts:18]:
```suggestion
 *     description: Basic information about a formula.
```

File: client/src/components/select-formula-view/SelectFormulaForm.tsx [hunk 6]
```diff
@@ -21,17 +21,19 @@ import StyledBox from './StyledBox';
...
-  const [codeSnippet, setCodeSnippet] = useState<any>('');
-  const [checkboxError, setCheckboxError] = useState<any>('');
-  const [formulaError, setFormulaError] = useState<any>('');
+  attainments, navigateToCourseView, navigateToAttributeSelection
+}: InferProps<typeof SelectFormulaForm.propTypes>): JSX.Element {
+  const [checkboxError, setCheckboxError]: State<string>  = useState('');
+  const [formulaError, setFormulaError]: State<string> = useState('');
   const [setSnackPack, messageInfo, setMessageInfo, alertOpen, setAlertOpen] = useSnackPackAlerts();
-
+  const [formulas, setFormulas]: State<Array<FormulaData> | null> = useState(null);
```
Comment (no-comma) [inline client/src/components/select-formula-view/SelectFormulaForm.tsx:36]:
```suggestion
  const [formulas, setFormulas]: State<Array<FormulaData>> = useState([]);
```

**Review by no-comma — CHANGES_REQUESTED** [review]

File: server/src/formulas/index.ts [hunk 7]
```diff
@@ -4,11 +4,12 @@
...
 import * as yup from 'yup';
 
-import { Formula, FormulaFunction, FormulaImplementation } from '../types/formulas';
+import { FormulaFunction, FormulaImplementation } from '../types/formulas';
+import { Formula } from 'aalto-grades-common/types';
 
 // The registry of formula implementations corresponding to their names, along
 // with a schema specifying what form their user parameters should take.
-const formulaImplementations: Map<Formula, FormulaImplementation> = new Map();
+export const formulaImplementations: Map<Formula, FormulaImplementation> = new Map();
```
[the remark this answers is not in the document]
Comment (jj-stigell) [reply server/src/formulas/index.ts]: removed the export and added new function to get relevant data and used the existing function to fetch implementation by id (somehow overlooked this initially)

File: client/src/components/formula-attributes-view/Attainment.tsx [hunk 8]
```diff
@@ -2,18 +2,21 @@
...
+import PropTypes, { InferProps } from 'prop-types';
 import Box from '@mui/material/Box';
 import Typography from '@mui/material/Typography';
 import TextField from '@mui/material/TextField';
 import formulaService from '../../services/formulas';
 
-function Attainment({ attainment, attributes, handleAttributeChange, attainmentIndex }) {
+function Attainment(
+  { attainment, attributes, handleAttributeChange, attainmentIndex }
+  : InferProps<typeof Attainment.propTypes>
```
[the remark this answers is not in the document]
Comment (jj-stigell) [reply client/src/components/formula-attributes-view/Attainment.tsx:13]: Totally forgot to bring this up in todays meeting. There seems to be multiple ways to declare types for the parameters in react components. Which one is best, I am not sure. Maybe discuss about this on fridays meeting?

File: server/src/formulas/weightedAverage.ts (hunk 4 above)
Comment (jj-stigell) [reply server/src/formulas/weightedAverage.ts]: removed

File: server/src/formulas/manual.ts [hunk 9]
```diff
@@ -23,8 +24,17 @@ async function manualGradeUnspecified(): Promise<CalculationResult> {
   };
 }
 
+const codeSnippet: string = 'placeholder';
+
+const name: string = 'Manual';
+
+const attributes: Array<string> = ['maxPoints', 'minRequiredPoints'];
```
[the remark this answers is not in the document]
Comment (jj-stigell) [reply server/src/formulas/manual.ts:31]: Mock data implementation did not even include the "manual" formula to the selection.  I added it to test how the dropdown menu and loading codesnippet works when switching between selected formulas. Should we remove manual from the list?

**Comment by jj-stigell** [comment]
All mock-data removed/moved to the test directory. Formula mock data temporarily moved to the formula service because few temporary functions use it (will not remove in this PR as it requires some planning how to efficiently fetch names and attributes for each formula without bombarding backend).

File: server/src/controllers/formula.ts [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+
+  res.status(HttpCode.Ok).json({
+    success: true,
+    data: {
+      formula: {
+        id: formulaId,
+        name: formula.name,
+        attributes: formula.attributes,
+        codeSnippet: formula.codeSnippet
+      }
```
Comment (no-comma) [inline server/src/controllers/formula.ts:44]: More of a general comment, but when we return data like this, if in the future we change the `FormulaPreview` interface, the compiler will not complain that this is wrong because the input of `json()` is `any`.

**Review by no-comma — APPROVED** [review]
LGTM, just a general comment.

File: server/src/controllers/formula.ts (hunk 10 above)
Comment (jj-stigell) [reply server/src/controllers/formula.ts]: good point, fixed
